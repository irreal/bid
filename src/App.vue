<template>
  <v-app id="inspire">
    <v-navigation-drawer
      dark
      v-if="isLoggedIn"
      v-model="drawer"
      class="blue-grey darken-2"
      fixed
      app
    >
      <v-toolbar flat class="transparent">
        <v-list class="pa-0">
          <v-list-tile avatar>
            <v-list-tile-avatar color="accent">
              <user-avatar class="white--text"></user-avatar>
            </v-list-tile-avatar>

            <v-list-tile-content>
              <v-list-tile-title>{{ username }}</v-list-tile-title>
            </v-list-tile-content>
          </v-list-tile>
        </v-list>
      </v-toolbar>
      <v-list dense>
        <v-list-tile @click="$router.push({ name: 'home' })">
          <v-list-tile-action>
            <v-icon>dashboard</v-icon>
          </v-list-tile-action>
          <v-list-tile-content>
            <v-list-tile-title>Dashboard</v-list-tile-title>
          </v-list-tile-content>
        </v-list-tile>
        <v-list-tile
          data-test="menu-button-sale-plans"
          class="menu-item-plans"
          @click="$router.push({ name: 'plans' })"
        >
          <v-list-tile-action>
            <v-icon>calendar_today</v-icon>
          </v-list-tile-action>
          <v-list-tile-content>
            <v-list-tile-title>Planovi Prodaje</v-list-tile-title>
          </v-list-tile-content>
        </v-list-tile>
        <v-list-tile
          class="menu-item-logout"
          @click="triggerNetlifyIdentityAction('logout')"
        >
          <v-list-tile-action>
            <v-icon>logout</v-icon>
          </v-list-tile-action>
          <v-list-tile-content>
            <v-list-tile-title>Izloguj se</v-list-tile-title>
          </v-list-tile-content>
        </v-list-tile>
      </v-list>
    </v-navigation-drawer>
    <v-toolbar color="primary" class="accent--text" fixed app dark>
      <v-toolbar-side-icon
        data-test="menu-toggle"
        v-if="isLoggedIn"
        @click.stop="drawer = !drawer"
      ></v-toolbar-side-icon>
      <v-toolbar-title>Darex BID</v-toolbar-title>
    </v-toolbar>
    <v-content>
      <v-fade-transition mode="out-in">
        <router-view></router-view>
      </v-fade-transition>
    </v-content>
    <v-footer v-if="isLoggedIn" color="primary" app>
      <span class="white--text">&copy; 2017</span>
    </v-footer>
  </v-app>
</template>
<style scoped lang="scss"></style>
<script>
import { mapGetters, mapActions } from "vuex";
import netlifyIdentity from "netlify-identity-widget";
import { EventBus } from "./event-bus.js";
import UserAvatar from "@/components/UserAvatar.vue";

netlifyIdentity.init({
  APIUrl: "https://www.darex.ml/.netlify/identity",
  logo: true
});

export default {
  name: "App",
  components: {
    UserAvatar
  },
  props: {
    source: String
  },
  computed: {
    ...mapGetters({
      isLoggedIn: "getUserStatus",
      user: "getUser"
    }),
    username() {
      return this.user
        ? this.user.user_metadata.full_name
        : "neulogovani korisnik";
    }
  },
  data: () => {
    return {
      nwid: netlifyIdentity,
      drawer: null
    };
  },
  mounted() {
    EventBus.$on("login", () => {
      this.triggerNetlifyIdentityAction("login");
    });
    EventBus.$on("logout", () => {
      this.triggerNetlifyIdentityAction("logout");
    });
    EventBus.$on("signup", () => {
      this.triggerNetlifyIdentityAction("signup");
    });
  },
  beforeDestroy() {
    EventBus.$off();
  },
  methods: {
    ...mapActions({
      updateUser: "updateUser"
    }),
    triggerNetlifyIdentityAction(action) {
      if (action == "login" || action == "signup") {
        netlifyIdentity.open(action);
        netlifyIdentity.on(action, user => {
          netlifyIdentity.close();
          this.updateUser({
            currentUser: user
          });
          if (user) {
            this.$router.push({ name: "home" });
          }
        });
      } else if (action == "logout") {
        this.updateUser({
          currentUser: null
        });
        netlifyIdentity.logout();
        this.$router.push({ name: "login" });
      }
    }
  }
};
</script>
