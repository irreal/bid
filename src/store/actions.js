export default {
  updateUser: ({ commit }, payload) => {
    commit("setUser", payload.currentUser);
  }
};
