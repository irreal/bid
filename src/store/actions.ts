import { ActionTree } from 'vuex';
import State from './state';

const actions: ActionTree<State, State> = {
    //   updateColor({ commit }: ActionContext<State, State>, newColor: string): void {
    //     commit(COLOR, newColor);
    //   },
    };

export default actions;
