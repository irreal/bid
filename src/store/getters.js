export default {
  getUserStatus: state => !!state.user,
  getUser: state => state.user
};
