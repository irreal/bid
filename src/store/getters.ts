import { GetterTree } from 'vuex';
import State from './state';
import Plan from '@/models/plan';

const getters: GetterTree<State, State> = {
    plans(state: State): Plan[] {
      return state.plans;
    },
    // colour(state: State): string {
    // 	return state.color;
    // },
  };

export default getters;
