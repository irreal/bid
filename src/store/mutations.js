export default {
  setUser: (state, currentUser) => {
    state.user = currentUser;
  }
};
