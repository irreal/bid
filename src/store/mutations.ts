import { MutationTree } from 'vuex';
import State from './state';

const mutations: MutationTree<State> = {
    removePlans(state: State) {
      state.plans = [];
    },
    // color(state: State, newColor: string): void {
    // 	state.color = newColor;
    // }
  };

export default mutations;
