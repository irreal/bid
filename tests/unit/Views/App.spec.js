import { mount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import App from "@/App";
import Vuetify from "vuetify";
import Vue from "vue";
import initialState from "@/store/state";
import actions from "@/store/actions";
import getters from "@/store/getters";
import mutations from "@/store/mutations";
import userFixture from "../fixtures/user";
import { EventBus } from "@/event-bus";
import netlifyIdentity from "netlify-identity-widget";

const localVue = createLocalVue();
localVue.use(Vuex);
Vue.use(Vuetify);

describe("App Root", () => {
  let state;

  const build = () => {
    const app = mount(App, {
      localVue,
      stubs: ["router-link", "router-view"],
      store: new Vuex.Store({ state, actions, getters, mutations })
    });

    return {
      app
    };
  };

  beforeEach(() => {
    state = { ...initialState };
  });
  it("renders the component", () => {
    // arrange
    const { app } = build();
    // assert
    expect(app.html()).toMatchSnapshot();
    app.destroy();
  });

  it("renders the component with a logged in user", () => {
    // arrange
    state.user = userFixture;
    const { app } = build();
    // assert
    expect(app.html()).toMatchSnapshot();
    app.destroy();
  });

  it("navigates to plans when menu item clicked", () => {
    state.user = userFixture;
    const { app } = build();
    app.vm.$router = { push: jest.fn() };
    app.find(".menu-item-plans").vm.$emit("click");

    expect(app.vm.$router.push).toBeCalledWith("plans");
    app.destroy();
  });

  it("responds to global events for netlify auth", () => {
    jest.mock("netlify-identity-widget");
    netlifyIdentity.open = jest.fn();
    netlifyIdentity.close = jest.fn();
    netlifyIdentity.logout = jest.fn();
    netlifyIdentity.on = (action, userCB) => {
      userCB(userFixture);
    };
    const { app } = build();
    app.vm.$router = { push: jest.fn() };

    EventBus.$emit("login");
    expect(netlifyIdentity.open).toBeCalledWith("login");
    expect(netlifyIdentity.close).toBeCalled();
    expect(app.vm.$router.push).toBeCalledWith({ name: "home" });

    EventBus.$emit("logout");
    expect(netlifyIdentity.logout).toBeCalled();
    expect(app.vm.$router.push).toBeCalledWith({ name: "login" });

    netlifyIdentity.close = jest.fn();
    app.vm.$router = { push: jest.fn() };
    EventBus.$emit("signup");
    expect(netlifyIdentity.open).toBeCalledWith("signup");
    expect(netlifyIdentity.close).toBeCalled();
    expect(app.vm.$router.push).toBeCalledWith({ name: "home" });

    netlifyIdentity.on = (action, userCB) => {
      userCB(null);
    };
    netlifyIdentity.close = jest.fn();
    app.vm.$router = { push: jest.fn() };
    EventBus.$emit("login");
    expect(netlifyIdentity.open).toBeCalledWith("login");
    expect(netlifyIdentity.close).toBeCalled();
    app.vm.$router.push = jest.fn();
    expect(app.vm.$router.push).not.toBeCalledWith({ name: "home" });

    app.vm.$router.push = jest.fn();
    netlifyIdentity.open = jest.fn();
    netlifyIdentity.close = jest.fn();
    netlifyIdentity.logout = jest.fn();
    netlifyIdentity.on = (action, userCB) => {
      userCB(userFixture);
    };
    app.vm.triggerNetlifyIdentityAction("unknown action");
    expect(app.vm.$router.push).not.toBeCalled();
    expect(netlifyIdentity.open).not.toBeCalled();
    expect(netlifyIdentity.close).not.toBeCalled();
    expect(netlifyIdentity.logout).not.toBeCalled();

    app.destroy();
  });

  it("reads username correctly", () => {
    const loggedOut = build();
    const username = loggedOut.app.vm.username;
    expect(username).toBe("neulogovani korisnik");
    loggedOut.app.destroy();

    state.user = userFixture;
    const loggedIn = build();
    const usernameLoggedIn = loggedIn.app.vm.username;
    expect(usernameLoggedIn).toBe(userFixture.user_metadata.full_name);
    loggedIn.app.destroy();
  });
});
