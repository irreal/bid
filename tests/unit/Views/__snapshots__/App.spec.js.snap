// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Root renders the component 1`] = `
<div data-app="true" class="application theme--light" id="inspire">
  <div class="application--wrap">
    <!---->
    <nav class="accent--text v-toolbar v-toolbar--fixed theme--dark primary" style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);">
      <div class="v-toolbar__content" style="height: 64px;">
        <!---->
        <div class="v-toolbar__title">Darex BID</div>
      </div>
    </nav>
    <main class="v-content" style="padding: 64px 0px 0px 0px;">
      <div class="v-content__wrap">
        <router-view-stub></router-view-stub>
      </div>
    </main>
    <!---->
  </div>
</div>
`;

exports[`App Root renders the component with a logged in user 1`] = `
<div data-app="true" class="application theme--light" id="inspire">
  <div class="application--wrap">
    <aside class="blue-grey darken-2 v-navigation-drawer v-navigation-drawer--close v-navigation-drawer--fixed v-navigation-drawer--is-mobile theme--dark" style="height: 100%; margin-top: 0px; transform: translateX(-300px); width: 300px;">
      <nav class="transparent v-toolbar elevation-0 theme--dark" style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);">
        <div class="v-toolbar__content" style="height: 64px;">
          <div role="list" class="v-list pa-0 theme--dark">
            <div role="listitem">
              <div class="v-list__tile v-list__tile--avatar theme--dark">
                <div class="v-list__tile__avatar">
                  <div class="v-avatar accent" style="height: 40px; width: 40px;"><span class="white--text">MS</span></div>
                </div>
                <div class="v-list__tile__content">
                  <div class="v-list__tile__title">Miloš Spasojević</div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </nav>
      <div role="list" class="v-list v-list--dense theme--dark">
        <div role="listitem"><a class="v-list__tile v-list__tile--link theme--dark">
            <div class="v-list__tile__action"><i aria-hidden="true" class="v-icon material-icons theme--dark">dashboard</i></div>
            <div class="v-list__tile__content">
              <div class="v-list__tile__title">Dashboard</div>
            </div>
          </a></div>
        <div role="listitem" class="menu-item-plans"><a class="v-list__tile v-list__tile--link theme--dark">
            <div class="v-list__tile__action"><i aria-hidden="true" class="v-icon material-icons theme--dark">calendar_today</i></div>
            <div class="v-list__tile__content">
              <div class="v-list__tile__title">Planovi Prodaje</div>
            </div>
          </a></div>
        <div role="listitem" class="menu-item-logout"><a class="v-list__tile v-list__tile--link theme--dark">
            <div class="v-list__tile__action"><i aria-hidden="true" class="v-icon material-icons theme--dark">logout</i></div>
            <div class="v-list__tile__content">
              <div class="v-list__tile__title">Izloguj se</div>
            </div>
          </a></div>
      </div>
      <div class="v-navigation-drawer__border"></div>
    </aside>
    <nav class="accent--text v-toolbar v-toolbar--fixed theme--dark primary" style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);">
      <div class="v-toolbar__content" style="height: 64px;"><button type="button" class="v-toolbar__side-icon v-btn v-btn--icon theme--dark">
          <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--dark">menu</i></div>
        </button>
        <div class="v-toolbar__title">Darex BID</div>
      </div>
    </nav>
    <main class="v-content" style="padding: 64px 0px 32px 0px;">
      <div class="v-content__wrap">
        <router-view-stub></router-view-stub>
      </div>
    </main>
    <footer class="v-footer v-footer--fixed theme--light primary" style="height: 32px;"><span class="white--text">© 2017</span></footer>
  </div>
</div>
`;
